### PR TITLE
Match seek by edge.

### PR DESCRIPTION
**Pull request by Shylock-Hg**

### CODE REVIEW

File: tests/tck/features/match/SeekByEdge.feature [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+      | "Shaquile O'Neal"    | "JaVale McGee"       | "Nuggets"       |
+      | "Shaquile O'Neal"    | "JaVale McGee"       | "Mavericks"     |
+
+  Scenario: seek by edge without index
+    When executing query:
+      """
+      MATCH (p1)-[:like]->(p2)
+      RETURN p1.name, id(p2)
+      """
+    Then a ExecutionError should be raised at runtime: Can't solve the start vids from the sentence: MATCH (p1)-[:like]->(p2) RETURN p1.name,id(p2)
```
Comment (czpmango) [inline tests/tck/features/match/SeekByEdge.feature:409]: Will we support this statement in the future? Neo4j is supported.

**Review by czpmango — APPROVED** [review]
Good job！

File: tests/tck/features/match/SeekByEdge.feature (hunk 1 above)
Comment (Shylock-Hg) [reply tests/tck/features/match/SeekByEdge.feature:409]: like has no index.

File: src/validator/MatchValidator.cpp [hunk 2]
```diff
@@ -243,7 +243,7 @@ Status MatchValidator::buildEdgeInfo(const MatchPath *path,
                     return Status::SemanticError("`%s': Unknown edge type", type->c_str());
                 }
                 edgeInfos[i].edgeTypes.emplace_back(etype.value());
-                edgeInfos[i].types.emplace_back(*type);
+                edgeInfos[i].types.emplace_back(type.get());
```
Comment (monadbobo) [inline src/validator/MatchValidator.cpp:246]: Is it possible for "type"(path->edge()->types()) to be released first here?

Comment (Shylock-Hg) [reply src/validator/MatchValidator.cpp:246]: The AST sentences is release when the query finished.

File: src/planner/match/LabelIndexSeek.cpp [hunk 3]
```diff
@@ -33,9 +33,25 @@ bool LabelIndexSeek::matchNode(NodeContext* nodeCtx) {
...
+        // TODO multiple tag index seek need the IndexScan support
+        return false;
+    }
+
+    edgeCtx->scanInfo.schemaIds = edge.edgeTypes;
+    edgeCtx->scanInfo.schemaNames = edge.types;
+
+    auto indexResult = pickEdgeIndex(edgeCtx);
+    if (!indexResult.ok()) {
+        return false;
```
Comment (laura-ding) [inline src/planner/match/LabelIndexSeek.cpp:49]: the status error msg need to print

File: src/planner/match/LabelIndexSeek.cpp [hunk 4]
```diff
@@ -65,9 +81,31 @@ StatusOr<SubPlan> LabelIndexSeek::transformNode(NodeContext* nodeCtx) {
     return plan;
 }
 
-StatusOr<SubPlan> LabelIndexSeek::transformEdge(EdgeContext*) {
-    // TODO
-    return Status::Error("TODO");
+StatusOr<SubPlan> LabelIndexSeek::transformEdge(EdgeContext* edgeCtx) {
+    SubPlan plan;
+    auto* matchClauseCtx = edgeCtx->matchClauseCtx;
+    DCHECK_EQ(edgeCtx->scanInfo.indexIds.size(), 1) << "Not supported multiple tag index seek.";
```
Comment (laura-ding) [inline src/planner/match/LabelIndexSeek.cpp:87]: modify `tag` to `edges`

File: src/planner/match/LabelIndexSeek.cpp [hunk 5]
```diff
@@ -98,5 +136,34 @@ StatusOr<SubPlan> LabelIndexSeek::transformEdge(EdgeContext*) {
...
+                if (candidateIndex == nullptr) {
+                    candidateIndex = index;
+                } else {
+                    candidateIndex = selectIndex(candidateIndex, index);
+                }
+            }
+        }
+        if (candidateIndex == nullptr) {
+            return Status::SemanticError("No valid index for label `%s'.",
+                                        edgeCtx->scanInfo.schemaNames[i]->c_str());
```
Comment (laura-ding) [inline src/planner/match/LabelIndexSeek.cpp:161]: alignment

File: src/planner/match/LabelIndexSeek.cpp [hunk 6]
```diff
@@ -33,9 +33,25 @@ bool LabelIndexSeek::matchNode(NodeContext* nodeCtx) {
     return true;
 }
 
-bool LabelIndexSeek::matchEdge(EdgeContext*) {
-    // TODO
-    return false;
+bool LabelIndexSeek::matchEdge(EdgeContext* edgeCtx) {
+    const auto &edge = *edgeCtx->info;
+    // require one edge at least
```
[the remark this answers is not in the document]
Comment (Shylock-Hg) [reply src/planner/match/LabelIndexSeek.cpp:38]: That's todo.

File: src/planner/match/LabelIndexSeek.cpp [hunk 7]
```diff
@@ -33,9 +33,25 @@ bool LabelIndexSeek::matchNode(NodeContext* nodeCtx) {
...
+    if (edge.edgeTypes.size() != 1) {
+        // TODO multiple tag index seek need the IndexScan support
```
[the remark this answers is not in the document]
Comment (Shylock-Hg) [reply src/planner/match/LabelIndexSeek.cpp]: Fixed.

File: src/planner/match/LabelIndexSeek.cpp (hunk 3 above)
Comment (Shylock-Hg) [reply src/planner/match/LabelIndexSeek.cpp:50]: Fixed

File: src/planner/match/LabelIndexSeek.cpp (hunk 4 above)
Comment (Shylock-Hg) [reply src/planner/match/LabelIndexSeek.cpp]: Fixed.

File: src/planner/match/LabelIndexSeek.cpp (hunk 5 above)
Comment (Shylock-Hg) [reply src/planner/match/LabelIndexSeek.cpp]: Fixed.

**Review by laura-ding — APPROVED** [review]

**Review by Aiee — APPROVED** [review]

**Review by yixinglu — COMMENTED** [review]
Does it only support to scan whole edge index? how to do about following scenarios if we have edge index on `like.likeness`:

```sql
MATCH (player)-[:like{likeness: 80}]-(player2) return player
```

If you have supported it, you can add some test cases in feature files.

**Comment by Shylock-Hg** [comment]
> Does it only support to scan whole edge index? how to do about following scenarios if we have edge index on `like.likeness`:
> 
> ```sql
> MATCH (player)-[:like{likeness: 80}]-(player2) return player
> ```
> 
> If you have supported it, you can add some test cases in feature files.

Had added.

File: src/planner/Query.h [hunk 8]
```diff
@@ -1112,6 +1112,43 @@ class UnionAllVersionVar final : public SingleInputNode {
         : SingleInputNode(qctx, Kind::kUnionAllVersionVar, input) {}
 };
 
+// merge columns to one
+// require the type of columns are same
+class ColumnsMerge final : public SingleInputNode {
```
Comment (yixinglu) [inline src/planner/Query.h:1117]: Why do you introduce this plan node ? Could project not do what you want ?

Comment (Shylock-Hg) [reply src/planner/Query.h:1117]: Project can't merge the different columns to one column.

**Review by yixinglu — COMMENTED** [review]
1. ColumnsMerge => Project{ a | b | c => [a, b, c]} => Unwind 
2. Use Gherkin `Examples` to merge `string` and `int` cases
3. Test variable length pattern match

**Comment by Shylock-Hg** [comment]
> 1. ColumnsMerge => Project{ a | b | c => [a, b, c]} => Unwind
> 2. Use Gherkin `Examples` to merge `string` and `int` cases
> 3. Test variable length pattern match

Fixed.

File: src/planner/match/PropIndexSeek.cpp [hunk 9]
```diff
@@ -13,13 +13,102 @@
 namespace nebula {
 namespace graph {
 bool PropIndexSeek::matchEdge(EdgeContext* edgeCtx) {
-    UNUSED(edgeCtx);
-    return false;
+    auto& edge = *edgeCtx->info;
+    if (edge.types.size() != 1 || edge.range != nullptr) {
+        // TODO multiple edge index seek need the IndexScan support
+        return false;
```
Comment (laura-ding) [inline src/planner/match/PropIndexSeek.cpp:19]: add log

File: src/planner/match/PropIndexSeek.cpp [hunk 10]
```diff
@@ -13,13 +13,102 @@
...
+        if (edge.props != nullptr && !edge.props->items().empty()) {
+            filter = MatchSolver::makeIndexFilter(*edge.types.back(),
+                                                   edge.props,
+                                                   matchClauseCtx->qctx,
+                                                   true);
+        }
+    }
+
+    if (filter == nullptr) {
+        return false;
```
Comment (laura-ding) [inline src/planner/match/PropIndexSeek.cpp:41]: ditto

File: src/planner/match/PropIndexSeek.cpp [hunk 11]
```diff
@@ -13,13 +13,102 @@
...
+    IQC iqctx;
+    iqctx.set_filter(Expression::encode(*edgeCtx->scanInfo.filter));
+    auto contexts = std::make_unique<std::vector<IQC>>();
+    contexts->emplace_back(std::move(iqctx));
+    auto columns = std::make_unique<std::vector<std::string>>();
+    std::vector<std::string> columnsName;
+    switch (edgeCtx->scanInfo.direction) {
+        case MatchEdge::Direction::OUT_EDGE:
+            columns->emplace_back(kSrc);
+            columnsName.emplace_back(kVid);
```
Comment (laura-ding) [inline src/planner/match/PropIndexSeek.cpp:67]: the columnsName why use the `kVid`

File: src/planner/match/PropIndexSeek.cpp [hunk 12]
```diff
@@ -13,13 +13,102 @@
...
+            break;
+        case MatchEdge::Direction::IN_EDGE:
+            columns->emplace_back(kDst);
+            columnsName.emplace_back(kVid);
```
Comment (laura-ding) [inline src/planner/match/PropIndexSeek.cpp:71]: ditto

File: src/planner/match/PropIndexSeek.cpp (hunk 11 above)
Comment (Shylock-Hg) [reply src/planner/match/PropIndexSeek.cpp:67]: Which better?

**Review by yixinglu — APPROVED** [review]

File: tests/tck/features/match/SeekByEdge.feature [hunk 13]
```diff
@@ -0,0 +1,937 @@
...
+  Background: Prepare space
+    Given a graph with space named "<space_name>"
+
+  Scenario: seek by edge index
+    When executing query:
+      """
+      MATCH (player)-[:serve]->(team)
+      RETURN player.name, team.name
+      """
+    Then the result should be, in any order:
```
Comment (czpmango) [inline tests/tck/features/match/SeekByEdge.feature:16]: Can you add filter or limit operators to make the number of query results less for easy verification?

Comment (czpmango) [reply tests/tck/features/match/SeekByEdge.feature:16]: NCK, mark resolved.

**Review by czpmango — APPROVED** [review]

**Pull request merged**